lib/utilities/retrieveData.js

```diff
@@ -47,24 +46,27 @@ export default function retrieveData(locations, retrieveDataFn, response, log) {
                 if (!responseDestroyed) {
                     _destroyResponse();
                 }
-                readable.emit('close');
+                readable.destroy();
                 return _next(responseErr);
             });
             // readable stream successfully consumed
-            readable.on('end', () => {
+            readable.once('end', () => {
+                readable.unpipe(response);
                 log.debug('readable stream end reached');
                 return _next();
             });
             // errors on server side with readable stream
-            readable.on('error', err => {
+            readable.once('error', err => {
                 log.error('error piping data from source');
+                _destroyResponse();
                 return _next(err);
             });
             return readable.pipe(response, { end: false });
         }), err => {
             if (err) {
                 log.debug('abort response due to client error', {
                     error: err.code, errMsg: err.message });
+		return _destroyResponse();
```

Comment:
I originally did not destroyResponse for all requests because for the situation where readable emits an error we have to send a 500 to the client instead of destroying the socket altogether.

Reply:
Oh. You're right. Got it. I'll remove that change.

Reply:
replying to self : If one of the reads frails from sproxyd, we cannot send back a 500 to the client : we already sent the headers with a 200, and content-length of the whole thing, so the only way to abort is for s3 to close the connection to the client when this happens.